src/coprocessor/dag/dag.rs

```diff
@@ -87,31 +84,93 @@ impl DAGContext {
                     }
                 }
-                Ok(None) => {
+                None => {
                     let mut resp = Response::new();
                     let mut sel_resp = SelectResponse::new();
                     sel_resp.set_chunks(RepeatedField::from_vec(chunks));
-                    let mut counts = Vec::with_capacity(4);
-                    self.exec.collect_output_counts(&mut counts);
-                    sel_resp.set_output_counts(counts);
+                    self.exec
+                        .collect_output_counts(sel_resp.mut_output_counts());
                     let data = box_try!(sel_resp.write_to_bytes());
                     resp.set_data(data);
                     return Ok(resp);
                 }
-                Err(e) => if let Error::Other(_) = e {
-                    let mut resp = Response::new();
-                    let mut sel_resp = SelectResponse::new();
-                    sel_resp.set_error(to_pb_error(&e));
-                    resp.set_data(box_try!(sel_resp.write_to_bytes()));
-                    resp.set_other_error(format!("{}", e));
-                    return Ok(resp);
-                } else {
-                    return Err(e);
-                },
             }
         }
     }
 
+    pub fn handle_streaming_request(
+        &mut self,
+        batch_row_limit: usize,
+    ) -> Result<(Option<Response>, bool)> {
+        let (mut record_cnt, mut finished) = (0, false);
+        let mut chunk = Chunk::new();
+        let mut start_key = None;
+        while record_cnt < batch_row_limit {
+            match self.exec.next()? {
+                Some(row) => {
+                    record_cnt += 1;
+                    if record_cnt == 1 {
+                        start_key = self.exec.take_last_key();
+                    }
+                    if self.has_aggr {
+                        chunk.mut_rows_data().extend_from_slice(&row.data.value);
+                    } else {
+                        let value = inflate_cols(&row, &self.columns, &self.output_offsets)?;
+                        chunk.mut_rows_data().extend_from_slice(&value);
+                    }
+                }
+                None => {
+                    finished = true;
+                    break;
+                }
+            }
+        }
+        if record_cnt > 0 {
+            let end_key = self.exec.take_last_key();
+            return self.make_stream_response(chunk, start_key, end_key)
+                .map(|r| (Some(r), finished));
+        }
+        Ok((None, true))
+    }
+
+    fn make_stream_response(
+        &mut self,
+        chunk: Chunk,
+        start_key: Option<Vec<u8>>,
+        end_key: Option<Vec<u8>>,
+    ) -> Result<Response> {
+        let mut s_resp = StreamResponse::new();
+        s_resp.set_encode_type(EncodeType::TypeDefault);
+        s_resp.set_data(box_try!(chunk.write_to_bytes()));
+        self.exec.collect_output_counts(s_resp.mut_output_counts());
```

Comment:
Should we consider `aggregation` or `topn` here @lamxTyler

Reply:
No.

Reply:
I think fix it in another PR is better.